npm/ipsm/ipsm.go

```diff
@@ -613,57 +605,59 @@ func (ipsMgr *IpsetManager) DestroyNpmIpsets() error {
 		return nil
 	}
 
-	entry := &ipsEntry{
-		operationFlag: util.IpsetFlushFlag,
-	}
-
+	destroyFailureCount := 0
 	for _, ipsetName := range ipsetLists {
-		entry := &ipsEntry{
+		flushEntry := &ipsEntry{
 			operationFlag: util.IpsetFlushFlag,
 			set:           ipsetName,
 		}
+		_, flushError := ipsMgr.run(flushEntry)
 
-		if _, err := ipsMgr.run(entry); err != nil {
-			metrics.SendErrorLogAndMetric(util.IpsmID, "{DestroyNpmIpsets} Error: failed to flush ipset %s", ipsetName)
+		deleteEntry := &ipsEntry{
+			operationFlag: util.IpsetDestroyFlag,
+			set:           ipsetName,
 		}
-	}
+		_, destroyError := ipsMgr.run(deleteEntry)
 
-	for _, ipsetName := range ipsetLists {
-		entry.operationFlag = util.IpsetDestroyFlag
-		entry.set = ipsetName
-		if _, err := ipsMgr.run(entry); err != nil {
+		if flushError != nil {
+			metrics.SendErrorLogAndMetric(util.IpsmID, "{DestroyNpmIpsets} Error: failed to flush ipset %s", ipsetName)
+		}
+		if destroyError != nil {
+			destroyFailureCount++
 			metrics.SendErrorLogAndMetric(util.IpsmID, "{DestroyNpmIpsets} Error: failed to destroy ipset %s", ipsetName)
 		}
+		if flushError == nil || destroyError == nil {
+			metrics.NumIPSetEntries.Add(float64(-metrics.GetIPSetInventory(ipsetName)))
+			metrics.RemoveFromIPSetInventory(ipsetName)
+		}
+	}
+
+	originalNumIPSets, numIPSetsError := promutil.GetValue(metrics.NumIPSets)
+	if originalNumIPSets > 0 && numIPSetsError == nil {
+		metrics.NumIPSets.Set(float64(destroyFailureCount))
+	} else {
+		metrics.NumIPSets.Set(0)
 	}
 
 	return nil
 }
 
-// Clean removes all the empty sets & lists under the namespace.
-func (ipsMgr *IpsetManager) Clean() error {
-	ipsMgr.Lock()
-	defer ipsMgr.Unlock()
-	for setName, set := range ipsMgr.setMap {
-		if len(set.elements) > 0 {
-			continue
-		}
-
-		if err := ipsMgr.deleteSet(setName); err != nil {
-			metrics.SendErrorLogAndMetric(util.IpsmID, "Error: failed to clean ipset")
-			return err
-		}
-	}
+func handleMetricCountsOnCreate() {
```

Comment:
My initial thought is the handleMetricCounts functions are specific to the implementation of IPSM; however, I'm noticing now that updates for NumIPSetEntries and IPSetInventory can be handled together nicely in a function like `metrics.IncIPSet(setName)` like your suggestion, and then we can remove those handle functions altogether since at most you'll be having two lines of duplicate code in IPSM for the deleteList and deleteSet functions:
```
metrics.FlushIPSet(setName)
metrics.NumIPSets.Dec()
```

As for not exposing the metrics, test utilities are in a separate folder right now (promutil), so the metrics do need to be exposed. 

@matmerr I think I remember you suggesting the test utility file go in a separate folder. Do you think  we could move this to the metrics package to encapsulate things?

Reply:
@huntergregory  it would agree with you and Junguk, having those functions in prom utils will be useful for next iteration of dataplane. metrics.FlushIPSet(setName) can be used in both windows and linux dataplane.